src/compiler/scala/tools/nsc/Global.scala

```diff
@@ -401,7 +405,7 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
 
     final def withCurrentUnit(unit: CompilationUnit)(task: => Unit): Unit = {
       if ((unit ne null) && unit.exists)
-        lastSeenSourceFile = unit.source
+        lastSeenSourceFile.set(unit.source)
```

Comment:
this is thread local - this would be expected to be the entry point for each task!